SlTree: separate expand/collapse and selection behaviour in 'single' mode

grncdr opened the pull request:
This changes the behaviour of sl-tree so that clicking on the expand/collapse icon will not select/deselect the item, only toggle it's expanded state.

See also: #1517 and discussion in #1159

grncdr committed:
Never select tree items when clicking the chevron

This changes the behaviour of sl-tree so that clicking on the expand/collapse icon will not select/deselect the item, only toggle it's expanded state.

claviska commented:
To continue the conversation from https://github.com/shoelace-style/shoelace/pull/1517#issuecomment-1680196881:

Ugh, this is a tough one. After using it awhile, the new behavior seems less intuitive than the original behavior, but of course reverting doesn't solve the original problem described in #1159.

I think your number three option is probably the best so users can cancel the necessary event(s) to achieve the behavior they want.

> Rewrite a fair bit of SlTree and SlTreeItem to dispatch cancellable events from user interaction before modifying the element properties/appearance. This would allow Shoelace consumers to to have precise control over the behaviour of trees and their items.

Otherwise, we'd need to pick the most common behaviors and toggle them with an attribute, which feels really weird and isn't something we're doing anywhere else in the library.

So this probably means more work, but hopefully it leads to a more flexible solution for everyone.

grncdr committed:
Refactor: inline SlTree.syncTreeItems

This was only called from 2 places, and they each had different
behaviour anyways.

grncdr committed:
SlTree: separate expand/collapse from selection

This makes 'multi' and 'single' mode consistent with each other, and
with native file managers.

grncdr commented:
> the new behavior seems less intuitive than the original behavior

I think the current state of this branch is in an uncanny valley/worst of both worlds. I'm going to push another change and maybe you can re-evaluate that.

grncdr commented:
> So this probably means more work, but hopefully it leads to a more flexible solution for everyone.

While I agree that it would be _nice_ to have that kind of fine control, I think you can kick the can a bit further down the road with improved default behaviour.

I haven't tested it, but it seems like users could add the old behaviour on top of this branch with an event listener like:

```ts
tree.addEventListener('sl-expand', (event) => {
  event.target.selected = !event.target.selected
})
```

claviska commented:
> What I've done is completely separated the selection and expand/collapse behaviour. My hypothesis is that most users expectations are going to be formed by the native file manager for their platform, and this is how Finder on OS X works. I don't have a Windows machine handy to test on, but I think explorer works the same way, clicking an expand/collapse control doesn't select an item, and selecting an item doesn't expand/collapse it.

This feels much better. It's a nice touch that `leaf` mode still expands/collapses on click, which I would expect. The single and multi-select feel very intuitive now.

grncdr commented:
Thanks for fixing the tests! I would have taken care of it… 😅

grncdr commented:
> @claviska I agree with the change made here, but it has one negative side-effect (IMO)... the user has to click on the open/close icon itself to get the toggle behavior. Previously, clicking on the tree item label would trigger open/close (which I think was due to selection, and which is why it no longer has that behavior).
> 
> What do you think about label click triggering open/close to make it easier UX?

@gap777 sorry for the slow response, but is setting `selection="leaf"` acceptable for your use case?

If you need to combine select+expand then you can always do something like this:

```
tree.addEventListener('sl-selection-change', (event) => {
  event.detail.selection[0].forEach(item => { item.expanded = true; })
})
```

gap777 commented:
The `selection='leaf'` seems acceptable. Thanks @grncdr !
